src/ReinforcementLearningZoo/src/algorithms/dqns/NFQ.jl

```diff
@@ -0,0 +1,80 @@
+export NFQ
+
+using Flux
+using Functors: @functor
+
+"""
+    NFQ{A<:AbstractApproximator, F, R} <: AbstractLearner
+    NFQ(approximator::A, num_iterations::Integer epochs::Integer, loss_function::F, batch_size::Integer, rng::R, γ::Float32) where {A<:AbstractApproximator, F, R}
+Neural Fitted Q-iteration as implemented in [1]
+
+# Keyword arguments
+- `approximator::AbstractApproximator` neural network
+- `num_iterations::Integer` number of value iteration iterations in FQI loop (i.e. the outer loop)
+- `epochs` number of epochs to train neural network per iteration
+- `loss_function::F` loss function of the NN
+- `sampler::BatchSampler{SARTS}` data sampler
+- `rng::R` random number generator
+- `γ::Float32` discount rate
+
+# References
+[1] Riedmiller, M. (2005). Neural Fitted Q Iteration – First Experiences with a Data Efficient Neural Reinforcement Learning Method. In: Gama, J., Camacho, R., Brazdil, P.B., Jorge, A.M., Torgo, L. (eds) Machine Learning: ECML 2005. ECML 2005. Lecture Notes in Computer Science(), vol 3720. Springer, Berlin, Heidelberg. https://doi.org/10.1007/11564096_32
+"""
+Base.@kwdef struct NFQ{A, R, F} <: AbstractLearner
+    action_space::AbstractVector
+    approximator::A
+    num_iterations::Integer = 20
+    epochs::Integer = 100
+    loss_function::F = mse
+    rng::R = Random.default_rng()
+    γ::Float32 = 0.9f0
+end
+
+# function NFQ(;
+#     action_space::AbstractVector,
+#     approximator::A,
+#     num_iterations::Integer = 20,
+#     epochs::Integer = 1000,
+#     loss_function::Any = mse,
+#     rng=Random.default_rng(),
+#     γ::Float32 = 0.9f0,
+#     ) where {A}
+#     NFQ(action_space, approximator, num_iterations, epochs, loss_function, rng, γ)
+# end
+
+# Copied from BasicDQN but sure whether it's appropriate
+@functor NFQ (approximator,)
+
+RLCore.forward(L::NFQ, s::AbstractArray) = RLCore.forward(L.approximator, s)
+
+function RLCore.forward(learner::NFQ, env::AbstractEnv)
+    as = action_space(env)
+    return vcat(repeat(state(env), inner=(1, length(as))), transpose(as)) |> x -> send_to_device(device(learner.approximator), x) |> x->RLCore.forward(learner, x) |> send_to_host |> vec 
+end
+
+# Avoid optimisation in the middle of an episode
+function RLBase.optimise!(::NFQ, ::AbstractStage, ::Trajectory) end
```

Comment:
Normally this is not necessary anymore with the latest master

Reply:
Right, NFQ is a learner not a policy. Then we should add 

`function RLBase.optimise!(::AbstractLearner, ::AbstractStage, ::Trajectory) end` 

in [learners.jl](https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/blob/main/src/ReinforcementLearningCore/src/policies/learners.jl). Mind doing that instead ? That way no unexpected behavior for future new learners.

Reply:
Resolved in 31b55b5 and b53c96b

Reply:
@HenriDeh  I see that you included the same fixes in #909. I'll wait till that is merged and then see about rolling back above commits and merging the updated main. After that NFQ should be good to go

Reply:
Okay, I'll wait for the review.